src/storage/src/storage_state/async_storage_worker.rs

```diff
@@ -167,20 +167,24 @@ where
         }
     }
 
-    // This cannot be instantiated earlier because `AsyncStorageWorker` then needs to be `+ Send +
-    // Sync` and capturing the timestamper in the Future makes it non-Send since it contains an Rc.
-    let mut timestamper = ReclockFollower::new(as_of);
-    timestamper.push_trace_batch(ReclockBatch {
-        updates: remap_updates,
-        upper: remap_upper.clone(),
-    });
+    remap_updates.sort_unstable_by(|a, b| a.1.cmp(&b.1));
 
+    let mut source_upper = MutableAntichain::new();
     let mut source_resume_uppers = BTreeMap::new();
     for (id, upper) in resume_uppers {
-        let source_upper = timestamper
-            .source_upper_at_frontier(upper.borrow())
-            .expect("enough data is loaded");
-        source_resume_uppers.insert(*id, source_upper);
+        let resume_upper = if PartialOrder::less_equal(upper, &as_of) {
+            Antichain::from_elem(Timestamp::minimum())
+        } else {
+            let idx = remap_updates.partition_point(|(_, t, _)| !upper.less_equal(t));
+            source_upper.clear();
+            source_upper.update_iter(
+                remap_updates[0..idx]
+                    .iter()
+                    .map(|(from_time, _, diff)| (from_time.clone(), *diff)),
+            );
+            source_upper.frontier().to_owned()
+        };
+        source_resume_uppers.insert(*id, resume_upper);
```

Comment:
we've lost the tests and comments associated with this logic due to removing the `source_upper_at_frontier` method -- can we add back some sort of explanation comment for this? Similar to the previous one?

```
    /// The conversion has the property that all messages that would be reclocked to times beyond
    /// the provided `IntoTime` frontier will be beyond the returned `FromTime` frontier. This can
    /// be used to compute a safe starting point to resume producing an `IntoTime` collection at a
    /// particular frontier.
```

Reply:
yep, done!